packages/core/crates/core-path/src/path.rs

```diff
@@ -1,82 +1,87 @@
+use crate::channel_graph::ChannelGraph;
 use crate::errors::PathError;
 use crate::errors::PathError::{ChannelNotOpened, InvalidPeer, LoopsNotAllowed, MissingChannel, PathNotValid};
 use crate::errors::Result;
 use core_crypto::types::OffchainPublicKey;
-use core_ethereum_db::traits::HoprCoreEthereumDbActions;
 use core_types::channels::ChannelStatus;
+use core_types::protocol::PeerAddressResolver;
+use futures::future::FutureExt;
+use futures::stream::FuturesOrdered;
+use futures::TryStreamExt;
 use libp2p_identity::PeerId;
+use std::collections::hash_map::RandomState;
+use std::collections::HashSet;
 use std::fmt::{Display, Formatter};
-use utils_log::warn;
+use std::hash::Hash;
 use utils_types::primitives::Address;
 use utils_types::traits::{PeerIdLike, ToHex};
 
-/// Represents a path for a packet.
-/// The type internally carries an information if the path has been already validated or not (since path validation
-/// is potentially an expensive operation).
-/// Path validation process checks if all the channels on the path exist and are in an `Open` state.
-#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
-#[derive(Debug, Clone)]
-pub struct Path {
-    hops: Vec<PeerId>,
-    valid: bool,
-}
+/// Base implementation of an abstract path.
+/// Must contain always at least a single entry.
+pub trait Path<N>: Display + Clone + Eq + PartialEq
+where
+    N: Copy + Clone + Eq + PartialEq + Hash,
+{
+    /// Individual hops in the path.
+    /// There must be always at least one hop.
+    fn hops(&self) -> &[N];
 
-#[cfg_attr(feature = "wasm", wasm_bindgen::prelude::wasm_bindgen)]
-impl Path {
-    /// Number of hops in the path.
-    pub fn length(&self) -> u32 {
-        self.hops.len() as u32
+    /// Shorthand for number of hops.
+    fn length(&self) -> usize {
+        self.hops().len()
     }
 
-    /// Determines with the path is valid.
-    pub fn valid(&self) -> bool {
-        !self.hops.is_empty() && self.valid
+    /// Gets the last hop
+    fn last_hop(&self) -> &N {
+        self.hops().last().expect("path is invalid")
```

Comment:
This panics and kills the application, but path input is a user input through the API. Would `Option<&N>` be acceptable?

Reply:
Both the `TransportPath` and `ChannelPath` are not allowed to be constructed empty, if constructed by the means of public constructors (or creator methods). So this is merely a `never happens` panic.